File: lingting-http/src/main/java/live/lingting/framework/http/download/HttpDownload.java
package live.lingting.framework.http.download;

import live.lingting.framework.download.MultipartDownload;
import live.lingting.framework.exception.DownloadException;
import live.lingting.framework.http.HttpClient;
import live.lingting.framework.http.HttpResponse;
import live.lingting.framework.http.header.HttpHeaders;
import live.lingting.framework.multipart.Part;
import live.lingting.framework.util.StreamUtils;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URI;
import java.net.http.HttpRequest;

/**
 * @author lingting 2023-12-20 16:43
 */
@SuppressWarnings("java:S1452")
public class HttpDownload extends MultipartDownload<HttpDownload> {

	protected final HttpClient client;

	protected final URI uri;

	public HttpDownload(HttpDownloadBuilder builder) throws IOException {
		super(builder);
		this.client = builder.client;
		this.uri = URI.create(url);
	}

	public static HttpDownloadBuilder builder(String url) {
		return new HttpDownloadBuilder(url);
	}

	public static HttpDownloadBuilder single(String url) {
		return new HttpDownloadBuilder(url).single();
	}

	public static HttpDownloadBuilder multi(String url) {
		return new HttpDownloadBuilder(url).multi();
	}

	public static HttpDownloadBuilder builder(URI url) {
		return new HttpDownloadBuilder(url);
	}

	public static HttpDownloadBuilder single(URI url) {
		return new HttpDownloadBuilder(url).single();
	}

	public static HttpDownloadBuilder multi(URI url) {
		return new HttpDownloadBuilder(url).multi();
	}

	void write(HttpRequest request, OutputStream output) throws IOException {
		HttpResponse response = client.request(request);

		if (!response.is2xx()) {
			throw new DownloadException(String.format("response status: %d", response.code()));
		}

		try (InputStream input = response.body()) {
			StreamUtils.write(input, output);
		}

	}

	@Override
	public long size() throws IOException {
		HttpRequest.Builder builder = HttpRequest.newBuilder().uri(uri).header("Accept-Encoding", "identity");
		HttpResponse response = client.request(builder.build());
		HttpHeaders headers = response.headers();
		return headers.contentLength();
	}

	@Override
	public InputStream download(Part part) throws Exception {
		HttpRequest.Builder builder = HttpRequest.newBuilder().GET().uri(uri);
		if (multi) {
			builder.header("Range", String.format("bytes=%d-%d", part.getStart(), part.getEnd()));
		}

		HttpRequest request = builder.build();
		HttpResponse response = client.request(request);

		if (!response.is2xx()) {
			throw new DownloadException(String.format("response status: %d", response.code()));
		}

		return response.body();
	}

}

File: lingting-http/src/main/kotlin/live/lingting/framework/http/download/HttpDownload.kt
package live.lingting.framework.http.download

import live.lingting.framework.data.DataSize
import live.lingting.framework.download.MultipartDownload
import live.lingting.framework.exception.DownloadException
import live.lingting.framework.http.HttpClient
import live.lingting.framework.http.HttpMethod
import live.lingting.framework.http.HttpRequest
import live.lingting.framework.multipart.Part
import live.lingting.framework.util.DataSizeUtils.bytes
import java.io.InputStream
import java.net.URI

/**
 * @author lingting 2023-12-20 16:43
 */
open class HttpDownload(builder: HttpDownloadBuilder) : MultipartDownload<HttpDownload>(builder) {

    companion object {
        @JvmStatic
        fun builder(url: String): HttpDownloadBuilder {
            return HttpDownloadBuilder(url)
        }

        @JvmStatic
        fun single(url: String): HttpDownloadBuilder {
            return HttpDownloadBuilder(url).single()
        }

        @JvmStatic
        fun multi(url: String): HttpDownloadBuilder {
            return HttpDownloadBuilder(url).multi()
        }

        @JvmStatic
        fun builder(url: URI): HttpDownloadBuilder {
            return HttpDownloadBuilder(url)
        }

        @JvmStatic
        fun single(url: URI): HttpDownloadBuilder {
            return HttpDownloadBuilder(url).single()
        }

        @JvmStatic
        fun multi(url: URI): HttpDownloadBuilder {
            return HttpDownloadBuilder(url).multi()
        }
    }

    protected val client: HttpClient = builder.client

    protected val uri: URI = URI.create(url)

    override fun size(): DataSize {
        val builder = HttpRequest.builder()
            .method(HttpMethod.HEAD)
            .url(uri)
            .header("Accept-Encoding", "identity")
        val response = client.request(builder.build())
        val headers = response.headers()
        return headers.contentLength().bytes
    }

    override fun download(part: Part): InputStream {
        val builder = HttpRequest.builder().get().url(uri)
        if (isMulti) {
            builder.header("Range", String.format("bytes=%d-%d", part.start.bytes, part.end.bytes))
        }

        val request = builder.build()
        val response = client.request(request)

        if (!response.is2xx) {
            throw DownloadException(String.format("response status: %d", response.code()))
        }

        return response.body()
    }

}
